### ISSUE TITLE
The future of webconfig

### ISSUE THREAD

**Issue by faho** [fish-shell/fish-shell#9461]
I've thought for quite a while that we need to do something about webconfig - our web-based configuration tool.

It isn't keeping up with new features, like our extended abbreviations - #9460.
It also makes implementing some features more annoying, like #9456 - which would be trivial in fish, but is awkward if you need to add it in two places and marshall the data through.

Another is that angularjs has gone EOL with no easy upgrade path in January 2022 - that's a year ago! (you may now feel old)

In short: It needs some real work put into it, and that isn't happening. Since 3.4.0 (released March 2022), it has received 16 commits (excluding those to themes and sample_prompts), most of them extremely small (5 just to the css). No substantial work on it has happened.

So, I propose one of two possibilities:

- We come up with a plan to port it off angular, and ideally get someone to maintain it and bring it up to speed
- We remove it - to be clear: the function `fish_config` and its CLI with the `theme` and `prompt` subcommands would stay, `fish_config browse`, the web-UI would go.

I realize that the latter is hard - webconfig is a flashy feature, and it has served fish well especially in the time that the CLI-version wasn't available. But now that we have `fish_config prompt` and `fish_config theme`, I believe that it doesn't really add a lot of value and it is time to let go.

**Comment by ridiculousfish** [comment]
I agree with your diagnosis and yet I hope to save web-UI `fish_config`. In my experience it's a powerful way to make users instantly comfortable with fish. Yes we do need to rewrite it from AngularJS, I have the stomach to do so, it's not that much code.

The observation about #9456 is good - `fish_config` does its magic through universal variables and we are trying to move away from those. It's a real problem if web-UI `fish_config` is an obstacle to new features.

Sorry I don't have any concrete conclusions at the moment, I need to think about it a bit.

**Comment by septatrix** [comment]
> Yes we do need to rewrite it from AngularJS, I have the stomach to do so, it's not that much code.

Do you already have a framework in mind? As it is a rather small app I would suggest Svelte or even lit-html as opposed to some of the heavier options like React, Vue, or Angular. I could also help with this if you want

**Comment by faho** [comment]
@septatrix I don't think we really care about which, specifically, just that it's light on dependencies - i.e. no need to run "npm" or ideally any tool whatsoever.

Really what we need here is basically a template engine - "add all themes in this spot".

**Comment by septatrix** [comment]
The requirements sound perfect for Alpine.js which I wanted to try out ever since I read about it (at least a year ago). Today I had a bit spare time so I worked on it and pushed the POC as a PR #9554. Feedback appreciated!

Obviously I did not yet tackle the technical backend aspects and instead tried to reimplement the status quo.

**Comment by siteshwar** [comment]
Chiming in as I did the AngularJS implementation almost a decade ago.

Related issues from past: 

https://github.com/fish-shell/fish-shell/pull/990
https://github.com/fish-shell/fish-shell/issues/1044

The problem with most of the JS frameworks is that they work like "pop culture". You will get a new contributor to port existing code to newer frameworks, but it's hard to keep such contributors for long periods as they move to other newer frameworks. If you choose to change the JS framework, you may remain stuck with it for another decade or so. It may be easiest to move to [Angular](https://angular.io/guide/upgrade), if you want to do minimal work and have to maintain it for a long period. AngularJS implementation worked well for almost a decade is a sign of success, even though it is considered little heavy weight than other frameworks. In comparison, there was an alternative implementation discussed with light weight [Knockout JS](https://github.com/knockout/knockout), but this framework seems to be out of maintenance with no clear migration path. When changing JS frameworks, I would be concerned much more about migration paths in future, than the nature of the framework itself.

In any case, webconfig should be kept and not be removed from this project.

**Comment by stelcodes** [comment]
I have not contributed to fish, but I'm a longtime user. I do like the webconfig editor in spirit but I have never seriously used it, even as a beginner shell user. I wouldn't mind seeing it go at all. The `fish_config` `theme` and `prompt` commands completely remove any need I had for a web GUI. What exactly is the point of keeping it around?

**Comment by ilyagr** [comment]
I find the new CLI interface much better than the web interface. Thank you for making it!

I tried using the web interface for themes, and the colors were all wrong for the 16-color themes at least (since it didn't know what palette my terminal uses).

Prompts are also easier with a CLI, since it's easier to try out in different directories.

**Comment by faho** [comment]
Okay, we've switched to alpine.

Frankly, it could still do with some love, but for now it'll work.
